packages/instantsearch.js/src/widgets/related-products/related-products.tsx

```diff
@@ -0,0 +1,222 @@
+        ? (data) => (
+            <TemplateComponent
+              {...renderState.templateProps}
+              templateKey="header"
+              rootTagName="fragment"
+              data={{
+                cssClasses: data.classNames,
+                recommendations: data.recommendations,
+              }}
+            />
+          )
+        : undefined
+    ) as RelatedProductsUiProps<Hit>['headerComponent'];
+
+    const itemComponent = (
+      templates.item
+        ? ({ item }) => {
+            return (
+              <TemplateComponent
+                {...renderState.templateProps}
+                templateKey="item"
+                rootTagName="fragment"
+                data={item}
+              />
+            );
+          }
+        : undefined
+    ) as RelatedProductsUiProps<Hit>['itemComponent'];
+
+    const fallbackComponent = (
+      templates.empty
+        ? () => (
+            <TemplateComponent
+              {...renderState.templateProps}
+              templateKey="empty"
+              rootTagName="fragment"
+              data={results}
+            />
+          )
+        : undefined
+    ) as RelatedProductsUiProps<Hit>['fallbackComponent'];
+
+    render(
+      <RelatedProducts
+        items={recommendations}
+        sendEvent={() => {}}
+        classNames={cssClasses}
+        headerComponent={headerComponent}
+        itemComponent={itemComponent}
+        fallbackComponent={fallbackComponent}
+        status={instantSearchInstance.status}
+      />,
+      containerNode
+    );
+  };
+}
+
+export type RelatedProductsCSSClasses = Partial<RecommendClassNames>;
+
+export type RelatedProductsTemplates = Partial<{
+  /**
+   * Template to use when there are no results.
+   */
+  empty: Template<RecommendResultItem>;
+
+  /**
+   * Template to use for the header of the widget.
+   */
+  header: Template<
+    Pick<
+      Parameters<
+        NonNullable<RelatedProductsUiProps<Hit>['headerComponent']>
+      >[0],
+      'recommendations'
+    > & { cssClasses: RecommendClassNames }
+  >;
+
+  /**
+   * Template to use for each result. This template will receive an object containing a single record.
+   */
+  item: Template<Hit>;
+}>;
+
+type RelatedProductsWidgetParams = {
+  /**
+   * CSS selector or `HTMLElement` to insert the widget into.
+   */
+  container: string | HTMLElement;
+
+  /**
+   * Templates to customize the widget.
+   */
+  templates?: RelatedProductsTemplates;
+
+  /**
+   * CSS classes to add to the widget elements.
+   */
+  cssClasses?: RelatedProductsCSSClasses;
+};
+
+export type RelatedProductsWidget = WidgetFactory<
+  RelatedProductsWidgetDescription & {
+    $$widgetType: 'ais.relatedProducts';
+  },
+  RelatedProductsConnectorParams,
+  RelatedProductsWidgetParams
+>;
+
+const relatedProducts: RelatedProductsWidget = function relatedProducts(
+  widgetParams
+) {
+  const {
+    container,
+    objectIDs,
+    maxRecommendations,
+    queryParameters,
```

Comment:
We should retrieve and forward `fallbackParameters` to the connector, as this model allows this parameter.

```suggestion
    queryParameters,
    fallbackParameters,
```